src/selectors/columnData.js

```diff
@@ -0,0 +1,148 @@
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule columnData
+ */
+
+'use strict';
+
+import {
+  extractPropsFromData,
+  extractPropsFromElement,
+  extractTemplatesFromData,
+  extractTemplatesFromElement,
+  getDefaultElementTemplates
+} from 'convertColumnElementsToData';
+import React from 'react';
+import forEach from 'lodash/forEach';
+import invariant from 'invariant';
+import isFunction from 'lodash/isFunction';
+import map from 'lodash/map';
+import shallowEqualSelector from 'shallowEqualSelector';
+import pick from 'lodash/pick';
+
+/**
+ * Converts React column / column group elements into props and cell rendering templates
+ */
+function convertColumnElementsToData(childComponents) {
+  const children = [];
+  React.Children.forEach(childComponents, (child) => {
+    if (child == null) {
+      return;
+    }
+    invariant(child.type.__TableColumnGroup__ || child.type.__TableColumn__,
+      'child type should be <FixedDataTableColumn /> or <FixedDataTableColumnGroup />');
+
+    children.push(child);
+  });
+
+  const elementTemplates = getDefaultElementTemplates();
+
+  const columnProps = [];
+  const hasGroupHeader = children.length && children[0].type.__TableColumnGroup__;
+  if (hasGroupHeader) {
+    const columnGroupProps = map(children, extractPropsFromElement);
+    forEach(children, (columnGroupElement, index) => {
+      elementTemplates.groupHeader.push(columnGroupElement.props.header);
+
+      React.Children.forEach(columnGroupElement.props.children, (child) => {
+        const column = extractPropsFromElement(child);
+        column.groupIdx = index;
+        columnProps.push(column);
+        extractTemplatesFromElement(elementTemplates, child);
+      });
+    });
+
+    return {
+      columnGroupProps,
+      columnProps,
+      elementTemplates,
+      useGroupHeader: true,
+    };
+  }
+
+  // Use a default column group
+  forEach(children, (child) => {
+    columnProps.push(extractPropsFromElement(child));
+    extractTemplatesFromElement(elementTemplates, child);
+  });
+  return {
+    columnGroupProps: [],
+    columnProps,
+    elementTemplates,
+    useGroupHeader: false,
+  };
+}
+
+function fixColumnPropsWithColumnGroupProps(columnProps, columnGroupProps) {
+  const columnGroups = [];
+  let columnsWithDefinedGroupIndex = 0;
+
+  forEach(columnProps, ({ groupIdx }) => {
+    if (groupIdx !== undefined && groupIdx !== null) {
+      columnGroups[groupIdx] = true;
+      columnsWithDefinedGroupIndex++;
+    }
+  });
+
+  invariant(columnGroups.length === (columnGroupProps || []).length,
+    'column group data must fully and only contain every column groups');
+
+  invariant(columnsWithDefinedGroupIndex === columnProps.length,
+    'group index if specified must be given for every column');
+
+  forEach(columnProps, (columnProp) => {
```

Comment:
There's a problem with grouped columns having fixed/fixedRight and the children not matching it.

Basically fixed/fixedRight works at separate levels for columnGroups/columns. This means fixed ColumnGroups will be fixed without the children columns being fixed, so there'll be misalignment.

For now, I've fixed it in the new API with this simple loop. 
@wcjordan, let me know if this should be added to the old one too. If so, we'll make  a separate PR (so that it won't be blocked by column virtualization)

Reply:
Hmm, I haven't heard too many complaints / confusion around this so I don't think it's necessary to add to the old setup, although it is nice to have.